### PR TITLE
QmlOverlay: Fix for issue use after delete

### PR DESCRIPTION
**Pull request by hyperlogic**
Sometimes, it was possible to set the qmlElement bounds AFTER the qmlElement itself was deleted on another thread.  To fix this we use a shared_ptr to hold the qmlElement, then use a weak_ptr in the setProperties lambda, to verify that the qmlElement still exists.

The script that was previously causing this code to fail is: "users.js".
Because, it creates a textSizeOverlay then immediately deletes it.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 64b56207f24c5aec17f9766f7bfe6424915db667 available at: https://gist.github.com/f59a968daeaa96752bf4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.dmg

**Comment by OmegaHeron** [comment]
Launched this repeatedly and didn't have crash on start or shutdown seen in production builds I suspect this is aimed at killing.

**Comment by jherico** [comment]
test this please

File: interface/src/ui/overlays/QmlOverlay.cpp [hunk 1]
```diff
@@ -22,20 +22,29 @@
 #include "Application.h"
 #include "text/FontFamilies.h"
 
-QmlOverlay::QmlOverlay(const QUrl& url) { 
+QmlOverlay::QmlOverlay(const QUrl& url) {
     buildQmlElement(url);
+    qDebug() << "AJT: url = " << url << (void*)QThread::currentThreadId();
```
Comment (jherico) [inline interface/src/ui/overlays/QmlOverlay.cpp]: remove debugging code before merge

File: interface/src/ui/overlays/QmlOverlay.cpp [hunk 2]
```diff
@@ -22,20 +22,29 @@
...
 }
 
 QmlOverlay::QmlOverlay(const QUrl& url, const QmlOverlay* textOverlay)
     : Overlay2D(textOverlay) {
+    qDebug() << "AJT: url = " << url << (void*)QThread::currentThreadId();
```
Comment (jherico) [inline interface/src/ui/overlays/QmlOverlay.cpp]: remove debugging code before merge

File: interface/src/ui/overlays/QmlOverlay.cpp [hunk 3]
```diff
@@ -22,20 +22,29 @@
...
     buildQmlElement(url);
 }
 
 void QmlOverlay::buildQmlElement(const QUrl& url) {
     auto offscreenUi = DependencyManager::get<OffscreenUi>();
+    auto threadId = QThread::currentThreadId();
```
Comment (jherico) [inline interface/src/ui/overlays/QmlOverlay.cpp]: remove debugging code before merge

File: interface/src/ui/overlays/QmlOverlay.cpp [hunk 4]
```diff
@@ -22,20 +22,29 @@
...
     offscreenUi->returnFromUiThread([=] {
         offscreenUi->load(url, [=](QQmlContext* context, QObject* object) {
-            _qmlElement = dynamic_cast<QQuickItem*>(object);
+            QQuickItem* rawPtr = dynamic_cast<QQuickItem*>(object);
+            // Create a shared ptr with a custom deleter lambda, that calls deleteLater
+            _qmlElement = std::shared_ptr<QQuickItem>(rawPtr, [=](QQuickItem* ptr) {
```
Comment (jherico) [inline interface/src/ui/overlays/QmlOverlay.cpp]: recommend `[]` rather than `[=]`.  There is nothing you need to capture here.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 64b56207f24c5aec17f9766f7bfe6424915db667 available at: https://gist.github.com/f59a968daeaa96752bf4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.dmg

**Comment by hyperlogic** [comment]
doh! sorry

**Comment by hifi-gustavo** [comment]
Build console output for commit: bea5ced3ebfed0620ccc2933cc640f2f151e7fe3 available at: https://gist.github.com/f59a968daeaa96752bf4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7002.dmg

**Pull request merged**